spark/src/main/java/org/apache/iceberg/RewriteManifestsAction.java

```diff
@@ -0,0 +1,490 @@
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT);
+    this.targetManifestSizeBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES,
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      fileIO = new HadoopFileIO(conf::value);
+    } else {
+      fileIO = table.io();
+    }
+  }
+
+  public RewriteManifestsAction rewriteIf(Predicate<ManifestFile> newPredicate) {
+    this.predicate = newPredicate;
+    return this;
+  }
+
+  public RewriteManifestsAction stagingLocation(String newStagingLocation) {
+    this.stagingLocation = newStagingLocation;
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsAction set(String property, String value) {
+    summary.put(property, value);
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsActionResult execute() {
+    Preconditions.checkArgument(stagingLocation != null, "Staging location must be set");
+
+    List<ManifestFile> matchingManifests = findMatchingManifests();
+    if (matchingManifests.isEmpty()) {
+      return null;
+    }
+
+    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
+
+    int parallelism = Math.min(matchingManifests.size(), defaultParallelism);
+    JavaRDD<ManifestFile> manifestRDD = sparkContext.parallelize(matchingManifests, parallelism);
+    Dataset<ManifestFile> manifestDS = spark.createDataset(manifestRDD.rdd(), manifestEncoder);
+    Dataset<Entry> manifestEntryDS = manifestDS.flatMap(toEntries(io, specs), entryEncoder);
+
+    try {
+      manifestEntryDS.cache();
+
+      long manifestEntrySizeBytes = computeManifestEntrySizeBytes(matchingManifests);
+      Map<Integer, List<PartitionMetadata>> metadataSizeSummary = computeMetadataSizeSummary(
+          manifestEntryDS,
+          manifestEntrySizeBytes);
+
+      Map<Integer, Map<StructLike, Integer>> bins = computeBins(metadataSizeSummary);
+
+      // the size of bins is estimated to be roughly targetManifestSizeBytes
+      // we allow the actual size of manifests to be 5% higher to avoid closing
+      // manifests near the end of bins if the estimation is not precise enough
+      // it is better to have slightly bigger manifests rather than manifests with a couple of entries
+      long manifestSizeBytes = (long) (1.05 * targetManifestSizeBytes);
+      List<ManifestFile> newManifests = manifestEntryDS
+          .groupByKey(toBin(bins), binEncoder)
+          .flatMapGroups(toManifest(io, manifestSizeBytes, stagingLocation, specs), manifestEncoder)
+          .collectAsList();
+
+      replaceManifests(matchingManifests, newManifests);
+
+      return new RewriteManifestsActionResult(matchingManifests, newManifests);
```

Comment:
If snapshot inheritance is not enabled, a rewritten version of `newManifests` will be committed. However, the action result will still return the staged manifests.

Reply:
Why does the action return something? Shouldn't this clean up any temporary files if it needs to?

Reply:
Initially, this action did not return anything. Then I started to think about returning a summary of what was done. In this case, it is a list of manifests that were replaced and added. Now, it seems a bit awkward. Maybe, we can drop this idea. Users can assign an id to the snapshot and play around with that to get some summary.

Reply:
I think it should handle temp cleanup, but returning something like stats is a good idea.